check commits are reimbursed in nu head

ffakenz opened the pull request:
This PR is a step into the right direction but there will be a follow-up one to address a gap in the `checkAbort`  where we should check that two identical commits can not be reimbursed by one output.

To check before merging:
* [x] CHANGELOG is up to date
* [x] ~Documentation is up to date~

abailly committed:
[wip] Move reimbursement check from Commit to Head

abailly committed:
[wip] Move reimbursement check from Commit to Head

v0d1ch committed:
Make it compile

v0d1ch committed:
Make it compile

v0d1ch committed:
Remove comments

v0d1ch reviewed (approved):
:ship:

ffakenz committed:
Update changelog

ffakenz committed:
minor

ch1bo commented on hydra-plutus/src/Hydra/Contract/Commit.hs:
```diff
@@ -1,6 +1,6 @@
 {-# LANGUAGE TemplateHaskell #-}
 {-# LANGUAGE TypeApplications #-}
-{-# OPTIONS_GHC -fno-specialize #-}
+{-# OPTIONS_GHC -fno-specialize -Wwarn #-}
```
Don't disable warnings as errors!?

ch1bo commented on hydra-plutus/src/Hydra/Contract/Commit.hs:
```diff
@@ -116,15 +114,7 @@ validator (_party, headScriptHash, commit) consumer ScriptContext{scriptContextT
-                          traceIfFalse "cannot find committed output" $
-                            preSerializedOutput `elem` (Builtins.serialiseData . toBuiltinData <$> txInfoOutputs txInfo)
-                    -- NOTE: In the Collectcom case the inclusion of the committed output 'commit' is
-                    -- delegated to the 'CollectCom' script who has more information to do it.
```
This NOTE was helpful. Similarly, a note for the abort case that the Head script will do the check that committed funds are reimbursed.

ch1bo commented on CHANGELOG.md:
```diff
@@ -14,6 +14,8 @@ changes.
 
 - HeadLogic Outcome is now being trace on every protocol step transition.
 
+- Check commits are reimbursed in nu head.
```
Let's call it `head validator`?

This is a `**BREAKING**` change as all other script changes.

ch1bo reviewed (changes requested):
Interestingly the `DropOneCommitOutput` mutation fails now. A possible solution could be to check that all PTs are burned in the `abortTx` (e.g. by the head validator).

pgrange committed:
FIX abort validator for commits

We should check the output of abort with the inputs of the transaction
and not the output.

pgrange committed:
FIX Abort validator

Not all the inputs we explore are commits, some may be intials so
we tolerate for the absence of datum as intial have no datum.

pgrange committed:
Secure collectCom validator tests

We explicitly check that one can not try to spend an initial in
a collectCom

pgrange committed:
Make the merge compile

pgrange committed:
Make the merge compile

pgrange committed:
FIX collectcom mutation

Head datum check has been removed from the commit validator

pgrange committed:
Address P.R. comments

pgrange committed:
FIX warnings

v0d1ch committed:
Don't use InitialState for checkCollectCom

v0d1ch committed:
Fix merge fuckup

v0d1ch committed:
Fix hydra-node tests

v0d1ch committed:
Add some haddock in v_commit

v0d1ch committed:
Add some haddock

ch1bo committed:
Detail changelog (also adding the ST in output)

ch1bo committed:
Document new mutation test for CollectCom

ch1bo committed:
Remove redundant wording in haddock and rename a variable

ch1bo committed:
Generate two commits to formulate attack on abort

Two commits with the same datum would only require one corresponding
output to have the validator pass.

v0d1ch committed:
Add new mutation to expose flaw in the checkAbort

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Abort.hs:
```diff
@@ -168,6 +169,12 @@ genAbortMutation (tx, _utxo) =
+        newInput <- arbitrary
+        -- XXX: Ideally we should need to modify the PT to simulate a proper new commit
+        -- FIXME: this shouldn't be green
+        pure $ AddInput newInput output
```
@v0d1ch and me added this to highlight that this PR is not completely solving the gap. We need to make sure the commits are completely reimbursed, i.e. not two identical commits can be reimbursed by one output. A follow-up PR will address this (and the corresponding FIXME in the script)
